### PR TITLE
Read The Internet Workshop Bounty

### PR DESCRIPTION
**Pull request by RafaelCenzano**

### CODE REVIEW

File: workshops/read_the_internet/README.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+# Reading the Internet
+We can read all the content on websites, but wouldn't it be cool if we could automatically read specific data from websites whenever you wanted? If this interested you, this workshop is for you. I will be showing you how to use webscraping in Python.
+**To Note:** not all webpages can be read using this technique, if data is loaded in using ajax or after the inital page request this method may not work.
+
+[Final Code](https://repl.it/@SavageCoder77/Read-The-Internet#main.py)
+[Live Result](https://Read-The-Internet.savagecoder77.repl.run)
+
+## Prerequisites
+You should have a basic understanding of **HTML** and **Python** or how to use variables and objects.
```
Comment (sampoder) [inline workshops/read_the_internet/README.md:16]:
```suggestion
You should have a basic understanding of **HTML** and **Python** and how to use variables and objects.
```

File: workshops/read_the_internet/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+## Let's do some setup
```
Comment (sampoder) [inline workshops/read_the_internet/README.md:18]:
```suggestion
## Setup
```

File: workshops/read_the_internet/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+![Lets start gif](https://cloud-p23nck0rj.vercel.app/1lets-start.gif)
+We will be using [repl.it](https://repl.it) an online code editor. This allows you to write and use this code from any computer!
```
Comment (sampoder) [inline workshops/read_the_internet/README.md:20]:
```suggestion
We will be using [repl.it](https://repl.it), an online code editor that allows you to write and run code from any computer!
```

File: workshops/read_the_internet/README.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+We will start by creating a Python 3 repl (You can easily create one by clicking this link https://repl.it/languages/python3).
+
+We now need to install a couple of external libraries. Follow the image below to find where to add the packages. Once in this menu you will need to install `beautifulsoup4`, `requests`, and `lxml`.  Documentation on the packages: [BS4](https://www.crummy.com/software/BeautifulSoup/bs4/doc/), [requests](https://requests.readthedocs.io/en/master/), and [LXML](https://lxml.de).
+![Repl.it packages menu](https://cloud-p5pdve0kf.vercel.app/0ezgif.com-video-to-gif.gif)
+
+
+#### A short description on the packages
```
Comment (sampoder) [inline workshops/read_the_internet/README.md:28]:
```suggestion
#### About the packages
```

**Review by sampoder — APPROVED** [review]
Yay! Looks good to me! Just a couple tiny edits andddd... could you please capitalize words like HTML, Python & COVID19

谢谢!

**Comment by RafaelCenzano** [comment]
Got all those things updated!

**Review by sampoder — APPROVED** [review]

File: workshops/read_the_internet/README.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
+---
+name: 'Reading the Internet'
+description: 'Use Python to read HTML and extract any information you can find!'
+author: '@RafaelCenzano'
+img: 'https://cloud-p5pdve0kf.vercel.app/2webscrape.jpg'
+---
+
+We can read all the content on websites, but wouldn't it be cool if we could automatically read specific data from websites whenever you wanted? If this interested you, this workshop is for you. I will be showing you how to use webscraping in Python.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:8]: This doesn't do the workshop justice. It's not just a tutorial on webscraping, you're actually building a tool that scrapes COVID case numbers from a website and displays them as text. This paragraph should be edited to more clearly show what you're going to build.

File: workshops/read_the_internet/README.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+*Note: not all webpages can be read using this technique. If data is loaded in using ajax or after the inital page request this method may not work.*
+
+[Final Result and Code](https://repl.it/@SavageCoder77/Read-The-Internet#main.py)
+
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:13]: Maybe also include a screenshot of the covid case numbers displayed in the terminal

File: workshops/read_the_internet/README.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+```
+
+Now that we have our BeautifulSoup object we can go all of the HTML! 
+
+### Working with the parsed HTML
+
+![HTML gif](https://cloud-5bmb4t5tp.vercel.app/2html.gif)
+
+In the next step, we'll be working with data from an HTML table. So, before, we continue, a quick mini-lesson on HTML tables:
+
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:75]: I think in this step it should be a lot more clear why you're suddenly learning about HTML tables. When I first read this it was a little jarring because I was like "wait what? i thought we were writing python. Why are we suddenly pausing to learn about HTML tables?"

File: workshops/read_the_internet/README.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        <td>Column 2 data</td>
+    </tr>
+</table>
+```
+
+A HTML table can be seen as a 2 dimensional array or list. The table contains rows (`tr` tags). Each row contains columns (`td` tags). Within the `td` tags is the data or text we can webscrape.
+
+### Python time!
+
+We're going to use BeautifulSoup to find an HTML table element. We want to find a specific table by ID `sortable_table_world` as seen in the picture. The code below finds the first table that matches that ID by using the `find` function and specifying the id as an attribute. The `attrs` is an optional argument and is a dictionary that can be set to any id, class, or HTML identifier.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:105]: During my first few reads, it wasn't totally clear to me what exactly this meant. I kinda get it now, but I'm still having trouble picturing why we're getting data from an HTML table or why this is necessary for webscraping, and I'm having trouble picturing what this step is actually doing. I think it'll be a lot more confusing to club members who are doing this workshop.

File: workshops/read_the_internet/README.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+![nCOVID table open with inspect element](https://cloud-p5pdve0kf.vercel.app/1inspect_element.png)
+
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:108]: You just finished explaining HTML tables, so you should have a transition to the next step here, before the code block.

File: workshops/read_the_internet/README.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+```python
+# get world data table
+table = soup.find('table', attrs={'id':'sortable_table_world'})
+```
+After this code runs we now have a new beautiful soup object called `table` that contains the table and child elements within that table.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:113]: Throughout the workshop, you keep referring the library as `beautiful soup`, but it's not a soup that's beautiful, it's a library called `BeautifulSoup`. I think you should go through and change all instances of `beautiful soup` to `BeautifulSoup`

File: workshops/read_the_internet/README.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+We will be using [repl.it](https://repl.it), an online code editor that allows you to write and run code from any computer!
+
+Spin up a Python3 repl by [clicking here](https://repl.it/languages/python3).
+
+Before we can start coding, we need to install a couple of external libraries. Follow the image below to find where to add the packages. Once in this menu you will need to install `beautifulsoup4`, `requests`, and `lxml`.
+
+![Repl.it packages menu](https://cloud-p5pdve0kf.vercel.app/0ezgif.com-video-to-gif.gif)
+
+#### What each package does
+`lxml` is a C package used by beautiful soup that parses the HTML, `beautifulsoup4` is for working with the parsed HTML, and `requests` is for requesting the webpage/html.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:29]: This is a good simple explanation for how each library applies to this workshop, but I think you should also include a little more information about what each library is. I still don't fully understand exactly what BeautifulSoup is—"working with the parsed HTML" doesn't mean much to me. Similar to `requests`: requests is a library that makes it super easy to make HTTP requests in Python, and in this workshop we're going to use it to "request" the HTML of a website.

You should make it super super clear, almost overexplaining what each library does, because understanding this is essential to understanding the rest of the workshop, and in its current state most club members will likely get lost if they don't understand what these libraries do.

File: workshops/read_the_internet/README.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+![nCOVID table open with inspect element](https://cloud-p5pdve0kf.vercel.app/1inspect_element.png)
+
+```python
+# get world data table
+table = soup.find('table', attrs={'id':'sortable_table_world'})
+```
+After this code runs we now have a new beautiful soup object called `table` that contains the table and child elements within that table.
+
+With `table` we will use the `findAll` function to get all the `tr` tags from the table. `tableRows` is a Python list with a beautiful soup object for each `tr` tag found in the table.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:115]: what is the `findAll` function? Where does it come from? Should be explained

File: workshops/read_the_internet/README.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+```python
+# get table rows
+tableRows = table.findAll('tr')
+```
+Now that we have all the rows in a Python list, we can iterate through the list and try to find the United States or you can input any Country you want.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:120]: awkward wording, this can be shortened or made a little clearer

File: workshops/read_the_internet/README.md [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+Before getting to the looping part though we will need to take a look at some of our outputs. This may happen to you when webscraping, the text has weird formatting. In these next steps I'm going to show the text from the website with the weird formatting and how we can use Python to remove that.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:122]: this paragraph is also kinda awkward wording and i don't feel like i'm fully understanding/picturing what i should be understanding/picturing after i read it. should be made clearer

File: workshops/read_the_internet/README.md [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+I will just use the second row and first column for this example
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:124]: now i'm lost. second row and first column of what?

File: workshops/read_the_internet/README.md [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+```python
+tableCols = tableRows[1].findAll('td')
+tableCol = tableCols[0]
+tableCol.text
+```
+This returns `'\r\n          TOTAL\r\n        '` with weird spacing and newline and indents. With Python we can use the `strip` function on the string and we will end with this code `tableCol.text.strip()` and this text `'TOTAL'`. Now that we got that handled lets get to reading the data!
+
+First we have a have a for each loop to loop through every row in the table. You make notice the `[1:]`, this was added to make the loop just a little more effcient because the first row at index `[0]` is just the table header with no data. Next we use an if statment to check if the current row's country is the United States (This can be replaced by any country). **Do Note** that we use the `.find()` function because we are only checking the first column and `.text[2:]` because the website has a weird astricks we need to remove. Once we find the country of our choice, we use `findAll()` to get all the columns in the row and print out some data.
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:132]: This is a very long paragraph with multiple explanations and instructions. It's very hard to follow. should be broken up

File: workshops/read_the_internet/README.md [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+The output at the time of writing this workshop is:
+```
+United States
+Confirmed Cases: 9,566,690
+New Cases: 28,845
+Critical Cases: 88,127
+```
+Scary Numbers! Hopefully at the time of reading this workshop things won't be too much worse :(
+
+Here is some example code for checking for multiple countries
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:151]: this confuses me because it's not worded as an instruction, so i'm not sure if i should copy/paste this code in order to continue the workshop or what exactly i should be doing

File: workshops/read_the_internet/README.md [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+    if row.find('td').text[2:].strip() in countries:
+        columns = row.findAll('td')
+        print(columns[0].text[2:].strip())
+        print(f'Confirmed Cases: {columns[1].text.strip()}')
+        print(f'New Cases: {columns[3].text.strip()}')
+        print(f'Critical Cases: {columns[5].text.strip()}\n')
+```
+
+### Thats it! You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!
+![Sponge bob done gif](https://cloud-5bmb4t5tp.vercel.app/0completed.gif)
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:164]: Click on "Preview" in this GitHub file and scroll down here. You may notice that the text appears next to the image instead of under it. This is because single new lines in markdown aren't interpreted as new lines. This is a problem I've been noticing throughout the workshop. It looks weird at first, but the proper formatting should be:

```
### That's it! You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!

![Sponge bob done gif](url)

following paragraph
```

That way it'll look correct on the website

File: workshops/read_the_internet/README.md [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+for row in tableRows[1:]:
+    if row.find('td').text[2:].strip() in countries:
+        columns = row.findAll('td')
+        print(columns[0].text[2:].strip())
+        print(f'Confirmed Cases: {columns[1].text.strip()}')
+        print(f'New Cases: {columns[3].text.strip()}')
+        print(f'Critical Cases: {columns[5].text.strip()}\n')
+```
+
+### Thats it! You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:163]: This is way too long for a heading. It shouldn't be a full sentence. Should probably be something like:

```
### That's it!

You can now get live COVID19 data but more importantly you are now primed with knowledge to webscrape anything!
```

Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:163]: Also this reads almost like a runon. May be useful to add some kind of punctuation after "COVID19 data"

File: workshops/read_the_internet/README.md [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+![Sponge bob done gif](https://cloud-5bmb4t5tp.vercel.app/0completed.gif)
+You can check out more advanced/in depth projects here. In this project you will be shown how to use selenium to webscrape that will allow you to webscrape pages like amazon or google that use ajax to load as you scroll (https://www.analyticsvidhya.com/blog/2020/08/web-scraping-selenium-with-python/). Check out this workshop by realpython that goes more in depth and links to other resources that can allow you to webscrape more complicated websites (https://realpython.com/beautiful-soup-web-scraper-python/).
```
Comment (MatthewStanciu) [inline workshops/read_the_internet/README.md:165]: This should not be dumped all in one paragraph one sentence after another. Instead, you should start a new section called "Hacking" and inside the section include each of these things in a list

**Review by MatthewStanciu — CHANGES_REQUESTED** [review]
Hey Raf, I'm glad to see this workshop finally submitted! 🙌 I think it still has a little bit of work before it's ready to merge, though. There are a lot of places throughout where I feel the grammar is weird or a topic isn't explained clearly enough, and there are multiple places where I got lost while going through the workshop because of that.

I think this workshop's biggest problem is that if I didn't know what webscraping actually was or how it worked, I would be totally lost by the end of the workshop. Webscraping is really cool! You're making web requests to a website to fetch its HTML contents, and then you're parsing the contents and working with the data you get in Python. That's really cool, but nothing like that is really explained anywhere in the workshop.

I've tried to point out as many of the places I thought were off as I could. I also made a few initial edits to the file so that you can get kinda a sense of what I'm thinking about.

If I can answer any questions, let me know on Slack. Also ping me on Slack once you've resubmitted so that I know to re-review (I don't check GitHub notifications very often)
